### ISSUE TITLE
pca components is incorrect shape

### ISSUE THREAD

**Issue by brentp** [mratsim/Arraymancer#356]
The docstring for pca seems to be what I expect, that given input shape of: `[Nb of observations, Nb of features]`, the shape of components should be: `[Nb of observations, Nb of components]`
however, this is not the case in the actual implementation.
 
here is code to reproduce:
```nim
import arraymancer

var x = newSeq[seq[float32]](100)
for r in x.mitems:
  r = newSeq[float32](32)

var xt = x.toTensor
echo xt.shape

var pr = xt.pca(2)

echo pr.results.shape
echo pr.components.shape

#doAssert pr.results.shape[0] == xt.shape[0]
# this fails because components.shape[0] matches xt.shape[1] 
doAssert pr.components.shape[0] == xt.shape[0]
```

**Comment by Vindaar** [comment]
I've never actually worked with PCA, but I have the impression the docstring may simply be wrong?

I fired up Python and checked an example with scikit-learn:
```python
import numpy as np
from sklearn.decomposition import PCA

x = np.array([[90, 60, 90],
              [90, 90, 30],
              [60, 60, 60],
              [60, 60, 90],
              [30, 30, 30]])
print(np.shape(x))
# (5, 3)
pca = PCA(n_components=2)
pca.fit(x)
y = pca.transform(x)
print(y) # equivalent to `pca.results`
# [[-34.37098481 -13.66927088]
#  [ -9.98345733  47.68820559]
#  [  3.93481353  -2.31599277]
#  [-14.69691716 -25.24923474]
#  [ 55.11654576  -6.45370719]]
print(np.shape(y))
# (5, 2)
print(pca.components_) # equivalent to `pca.components`
# [[-0.65580225 -0.4291978  -0.62105769]
#  [ 0.3859988   0.51636642 -0.7644414 ]]
print(np.shape(pca.components_))
# (2, 3)
```

and the equivalent in arraymancer:
```nim
import arraymancer

let xt = [[90'f64, 60, 90],
          [90'f64, 90, 30],
          [60'f64, 60, 60],
          [60'f64, 60, 90],
          [30'f64, 30, 30]].toTensor
echo xt.shape
# [5, 3]
var pr = xt.pca(2)

echo pr.results
# Tensor[system.float64] of shape [5, 2] of type "float64" on backend "Cpu"
# |-34.37098480726787     13.66927088271195|
# |-9.983457328239837     -47.68820559298736|
# |3.934813529880883      2.315992772328602|
# |-14.69691715786346     25.24923474435496|
# |55.11654576349029      6.453707193591847|
echo pr.results.shape
# [5, 2]

echo pr.components
# Tensor[system.float64] of shape [3, 2] of type "float64" on backend "Cpu"
# |-0.6558022549801472    -0.3859987953881002|
# |-0.4291977965486883    -0.5163664177215532|
# |-0.6210576895914781    0.7644413990675453|
echo pr.components.shape
# [3, 2]
```
So while the output of `components` of scikit-learn is transposed compared to arraymancers (and the sign is flipped in arraymancer's column 1), the dimensions are `[Nb of features, Nb of components]` and *not* `[Nb of observations, Nb of components]` as stated in the docstring.

**Comment by brentp** [comment]
thanks for looking into this. I'll have a look next week. The other thing I noticed is that I couldn't use the result.components in the call to `pca` that does the projection (shape error), but maybe I just needed to transpose.

**Comment by brentp** [comment]
actually, this is working for me, so I think you're right, it's just the docstring that needs correcting:
```
import arraymancer

var x = newSeq[seq[float32]](100)
for r in x.mitems:
  r = newSeq[float32](320)

var xt = x.toTensor 
# [100, 320]
echo xt.shape

var pr = xt.pca(2)

# [100, 2]
echo "pr.results.shape:", pr.results.shape
# [320, 2]
echo "pr.components.shape:", pr.components.shape

# [100, 2]
echo xt.pca(pr.components).shape
```

**Comment by brentp** [comment]
this should probably be a separate issue, but any chance of a randomzed pca? sklearn uses this: https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/extmath.py#L230

when using solver=randomized and that finishes in ~5 seconds for something that takes arraymancer 250 seconds (shape is [2504,  16000])

**Comment by mratsim** [comment]
Reopening to fix either the docstrings or the order of the components.

I actually didn't check the components order versus Scikit-learn and have no tests to enforce a specific order.

The current components order was chosen because it multiplied naturally with mean to obtained the pca:

https://github.com/mratsim/Arraymancer/blob/18b596bcf1a886e185becdda3342458cfa7b1c35/src/ml/dimensionality_reduction/pca.nim#L30

And you could pass those directly as well to the overload that accepts principal axes

https://github.com/mratsim/Arraymancer/blob/18b596bcf1a886e185becdda3342458cfa7b1c35/src/ml/dimensionality_reduction/pca.nim#L32-L41

Regarding RandomizedPCA, I'm not against it but it's not a priority for me, I think implementing PCA via SVD instead of eigenvectors like today would help a lot as well.

**Comment by brentp** [comment]
you are right, even without using randomized pca, sklearn can do the PCA.fit().transform() in 1.6 seconds.

**Comment by mratsim** [comment]
Shape issues in scikit-learn: https://stackoverflow.com/questions/38276047/scikit-learn-pca-fit-transform-shape-is-inconsistent-n-samples-m-attributes

This should also be tested and mentionned for the docstring update

**Comment by mratsim** [comment]
I'm consistent with R as well:
![DeepinScreenshot_select-area_20190930135053](https://user-images.githubusercontent.com/22738317/65876458-8bfab980-e389-11e9-9c2e-d49c9e1091bd.png)

**Comment by mratsim** [comment]
doc fixed in #384 - https://github.com/mratsim/Arraymancer/blob/45a6674718483b037bf01aedddab27d977c7a179/src/ml/dimensionality_reduction/pca.nim#L9-L43
